### ISSUE TITLE
Actual memory usage appears much larger than psi4.set_memory()

### ISSUE THREAD

**Issue by slochower** [psi4/psi4#1985]
I see memory usage significantly above `psi4.set_memory()` limit. I do see the note in the documentation that memory usage might be _slightly_ higher, but if I even request within 20% of a node's total memory (requesting 300 out of ~376 GB), I risk crashing the entire node. Even now, running with `psi4.set_memory("200 GB")`, I see persistent VIRT usage of 259 GB and RES 256 GB. I know that determining the _actual_ amount of [memory consumed by a process is tricky](https://stackoverflow.com/questions/131303/how-to-measure-actual-memory-usage-of-an-application-or-process), but this is important to me to prevent the scheduler from killing my jobs. After some troubleshooting, I am requesting 300 GB from the scheduler, but 200 GB from psi4, and end up with ~260 GB used. Is this expected? 

Relatedly, but separate, is there a way to estimate what the scratch file size may be? I was planning to use node-local storage until I found out my scratch files are 1.5+ TB. Any rule of thumb way of estimating this ahead of time (number of atoms × level of theory × some magic thing?)

**Comment by CDSherrill** [comment]
You may have found a memory allocation/usage bug.  What kind of
computation is this?

Unfortunately we don't really have a way to estimate scratch usage at
present. It's a feature we'd like to add, but it needs to be added for
every module separately, so is a somewhat daunting task.  In principle, can
be known ahead of time based on number of basis functions, number of atoms,
etc., but is different for every method (and sometimes depends on the
particular algorithm used within that method).

**Comment by hokru** [comment]
Would be good to know the psi4 version and the input options used.

I know that `scf_type pk` can get very greedy for larger calculations.

**Comment by slochower** [comment]
I'm running a single point energy calculation with `psi4.energy("ccsd(t)/aug-cc-pvdz")`. 

The input is simple. Just:

```
import psi4
from psi4 import *

psi4.core.set_output_file('xxx.dat', False)

psi4.set_memory('200 GB')
psi4_io = core.IOManager.shared_object()
psi4_io.set_default_path("xxx")
psi4.set_num_threads(24)
```
followed by the molecule specification with `psi4.geometry()` and then
```
psi4.energy("ccsd(t)/aug-cc-pvdz")
```

(I know this will take a lot of memory and disk space, but I believe the memory limit should stand.)

- psi4 version 1.3.2 in a new python 3.7 conda environment
- RHEL 7.2

**Comment by hokru** [comment]
Yes the memory limit should stand. Ideally to perhaps less than 10% excess. 

Do you know at which point in the calculation this excessive request of memory stems from?
And how large is the calculations? (number of basis functions and if you know occupied/virtual orbitals?).
If you are willing to share, uploading the output would help. You can remove the molecule information if you wish.

frozen core approximation (off by default) and density fitting will probably go a long way to make such a large calculation feasible.

**Comment by slochower** [comment]
I'm not familiar enough with the output to know at what point in the calculation the memory excess occurs, but it's not just a single point in time—the memory stays above 250 GB for quite some time. Attached is the output file with the molecule coordinates scrubbed along with a screenshot during running.

![memory](https://user-images.githubusercontent.com/5393995/91202565-d7f7d080-e6b6-11ea-8c10-cb83ca8420f9.png)
[output.dat](https://github.com/psi4/psi4/files/5125101/tmp.txt)

**Comment by hokru** [comment]
Thanks! This *is* a large calculation (>900bfs) and I fear the conventional CCSD(T) will struggle with disk space and disk I/O once the current slow task of writing the 2e-ints to disk (again) is done.
I have an idea what is happening but need to test it.

**Comment by hokru** [comment]
I proposed fixes that should keep the SCF memory in check in the future. Thanks again for reporting.

An alternative for psi4.1.3.2 is `scf_type direct` if you want to keep conventional integrals.
However, I recommend DF-CCSD(T), perhaps even DF-FNO-CCSD(T). There is no practical loss in accuracy, especially with a small basis like aug-cc-pVDZ. (`cc_type df`).
And of course frozen core approximation (`frozen_core true`).

**Comment by slochower** [comment]
Thanks, @hokru. I haven't had time to test the particular fix on my job yet, but I appreciate the effort and also the tips :) Feel free to close this issue if you feel like the PR is sufficient.
